Swift's testRefresh(), testAutorefresh() and testRealm() occasionally fail

By leaking realms: `RealmSwift/Tests/TestCase.swift:44: error: -[RealmSwift_Tests.RealmTests testRefresh] : XCTAssertNil failed: "Optional(<RLMRealm: 0x103f1d110>)"`. See https://ci.realm.io/job/objc_pr/1009/configuration=Debug,target=osx-swift/console.

--- a/RealmSwift/Tests/ObjectTests.swift
+++ b/RealmSwift/Tests/ObjectTests.swift
@@ -40,13 +40,11 @@ class ObjectTests: TestCase {
         XCTAssertNotNil(persisted.realm)
         XCTAssertEqual(realm, persisted.realm!)
 
-        let queue = dispatch_queue_create("background", nil)
-        dispatch_async(queue) {
+        dispatchSyncNewThread {
             autoreleasepool {
                 XCTAssertNotEqual(Realm(), persisted.realm!)
             }
         }
-        dispatch_sync(queue) {}
     }
 
     func testObjectSchema() {

--- a/RealmSwift/Tests/ObjectTests.swift
+++ b/RealmSwift/Tests/ObjectTests.swift
@@ -40,11 +40,13 @@ class ObjectTests: TestCase {
         XCTAssertNotNil(persisted.realm)
         XCTAssertEqual(realm, persisted.realm!)
 
-        let queue = dispatch_queue_create("background", DISPATCH_QUEUE_SERIAL)
+        let queue = dispatch_queue_create("background", nil)
         dispatch_async(queue) {
-            XCTAssertNotEqual(Realm(), persisted.realm!)
+            autoreleasepool {
+                XCTAssertNotEqual(Realm(), persisted.realm!)
+            }
         }
-        dispatch_sync(queue, {})
+        dispatch_sync(queue) {}
     }
 
     func testObjectSchema() {

--- a/RealmSwift/Tests/PerformanceTests.swift
+++ b/RealmSwift/Tests/PerformanceTests.swift
@@ -336,11 +336,9 @@ class SwiftPerformanceTests: TestCase {
 
     func testRealmCreationCached() {
         var realm: Realm!
-        let queue = dispatch_queue_create("test queue", nil)
-        dispatch_async(queue) {
+        dispatchSyncNewThread {
             realm = self.realmWithTestPath()
         }
-        dispatch_sync(queue) {}
 
         measureBlock {
             for _ in 0..<250 {

--- a/RealmSwift/Tests/RealmTests.swift
+++ b/RealmSwift/Tests/RealmTests.swift
@@ -395,15 +395,12 @@ class RealmTests: TestCase {
             notificationFired.fulfill()
         }
 
-        let queue = dispatch_queue_create("background", nil)
-        dispatch_async(queue) {
+        dispatchSyncNewThread {
             Realm().write {
                 Realm().create(SwiftStringObject.self, value: ["string"])
-                return
             }
         }
-        waitForExpectationsWithTimeout(2, handler: nil)
-        dispatch_sync(queue) {}
+        waitForExpectationsWithTimeout(1, handler: nil)
         Realm().removeNotification(token)
 
         // get object
@@ -427,15 +424,13 @@ class RealmTests: TestCase {
         let results = realm.objects(SwiftStringObject)
         XCTAssertEqual(results.count, Int(0), "There should be 1 object of type StringObject")
 
-        let queue = dispatch_queue_create("background", nil)
-        dispatch_async(queue) {
+        dispatchSyncNewThread {
             Realm().write {
                 Realm().create(SwiftStringObject.self, value: ["string"])
                 return
             }
         }
-        waitForExpectationsWithTimeout(2, handler: nil)
-        dispatch_sync(queue) {}
+        waitForExpectationsWithTimeout(1, handler: nil)
         realm.removeNotification(token)
 
         XCTAssertEqual(results.count, Int(0), "There should be 1 object of type StringObject")
@@ -492,14 +487,9 @@ class RealmTests: TestCase {
         let testRealm = realmWithTestPath()
         XCTAssertFalse(realm == testRealm)
 
-        let fired = expectationWithDescription("fired")
-        let queue = dispatch_queue_create("background", nil)
-        dispatch_async(queue) {
+        dispatchSyncNewThread {
             let otherThreadRealm = Realm()
             XCTAssertFalse(realm == otherThreadRealm)
-            fired.fulfill()
         }
-        waitForExpectationsWithTimeout(2, handler: nil)
-        dispatch_sync(queue) {}
     }
 }

--- a/RealmSwift/Tests/RealmTests.swift
+++ b/RealmSwift/Tests/RealmTests.swift
@@ -395,13 +395,15 @@ class RealmTests: TestCase {
             notificationFired.fulfill()
         }
 
-        dispatch_async(dispatch_queue_create("background", nil)) {
+        let queue = dispatch_queue_create("background", nil)
+        dispatch_async(queue) {
             Realm().write {
                 Realm().create(SwiftStringObject.self, value: ["string"])
                 return
             }
         }
         waitForExpectationsWithTimeout(2, handler: nil)
+        dispatch_sync(queue) {}
         Realm().removeNotification(token)
 
         // get object
@@ -425,13 +427,15 @@ class RealmTests: TestCase {
         let results = realm.objects(SwiftStringObject)
         XCTAssertEqual(results.count, Int(0), "There should be 1 object of type StringObject")
 
-        dispatch_async(dispatch_queue_create("background", nil)) {
+        let queue = dispatch_queue_create("background", nil)
+        dispatch_async(queue) {
             Realm().write {
                 Realm().create(SwiftStringObject.self, value: ["string"])
                 return
             }
         }
         waitForExpectationsWithTimeout(2, handler: nil)
+        dispatch_sync(queue) {}
         realm.removeNotification(token)
 
         XCTAssertEqual(results.count, Int(0), "There should be 1 object of type StringObject")
@@ -489,11 +493,13 @@ class RealmTests: TestCase {
         XCTAssertFalse(realm == testRealm)
 
         let fired = expectationWithDescription("fired")
-        dispatch_async(dispatch_queue_create("background", nil)) {
+        let queue = dispatch_queue_create("background", nil)
+        dispatch_async(queue) {
             let otherThreadRealm = Realm()
             XCTAssertFalse(realm == otherThreadRealm)
             fired.fulfill()
         }
         waitForExpectationsWithTimeout(2, handler: nil)
+        dispatch_sync(queue) {}
     }
 }

--- a/RealmSwift/Tests/TestCase.swift
+++ b/RealmSwift/Tests/TestCase.swift
@@ -48,6 +48,12 @@ class TestCase: XCTestCase {
         RLMRealm.resetRealmState()
     }
 
+    func dispatchSyncNewThread(block: dispatch_block_t) {
+        let queue = dispatch_queue_create("background", nil)
+        dispatch_async(queue, block)
+        dispatch_sync(queue) {}
+    }
+
     func assertThrows<T>(@autoclosure(escaping) block: () -> T, _ message: String? = nil, fileName: String = __FILE__, lineNumber: UInt = __LINE__) {
         exceptionThrown = true
         RLMAssertThrows(self, { _ = block() } as dispatch_block_t, message, fileName, lineNumber)

